File: Erlang/0012.erl
-spec int_to_roman(integer()) -> unicode:unicode_binary().
int_to_roman(Num) when Num >= 1, Num =< 3999 ->
    int_to_roman(Num, [
        {1000, <<"M">>}, {900, <<"CM">>}, {500, <<"D">>}, {400, <<"CD">>},
        {100, <<"C">>}, {90, <<"XC">>}, {50, <<"L">>}, {40, <<"XL">>},
        {10, <<"X">>}, {9, <<"IX">>}, {5, <<"V">>}, {4, <<"IV">>}, {1, <<"I">>}
    ]).

int_to_roman(0, _) -> <<>>;
int_to_roman(Num, [{Value, Symbol} | Rest]) when Num >= Value ->
    <<Symbol/binary, (int_to_roman(Num - Value, [{Value, Symbol} | Rest]))/binary>>;
int_to_roman(Num, [_ | Rest]) ->
    int_to_roman(Num, Rest).

File: Elixir/0012.exs
defmodule Solution do
  @spec int_to_roman(integer) :: String.t()
  def int_to_roman(num) do
    values = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
    symbols = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"]

    int_to_roman(num, values, symbols, "")
  end

  defp int_to_roman(0, _values, _symbols, result), do: result

  defp int_to_roman(num, [value | rest_values], [symbol | rest_symbols], result) do
    if num >= value do
      int_to_roman(num - value, [value | rest_values], [symbol | rest_symbols], result <> symbol)
    else
      int_to_roman(num, rest_values, rest_symbols, result)
    end
  end
end
